src/System.Management.Automation/engine/parser/tokenizer.cs

```diff
@@ -3326,204 +3370,271 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, Number
                             {
-                                result = (ushort)(us * (ushort)multiplier);
+                                result = dm;
                                 return true;
                             }
 
-                            result = null;
-                            return false;
-                        case NumberSuffixFlags.UnsignedByte:
-                            // If multiplier is negative or greater than 1, we can assume it will fail since the
-                            // minimum multiplier is 1024 (already exceeds byte.MaxValue), and byte is unsigned
-                            if (multiplier == 1 && byte.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out byte b))
+                            break;
+                        case NumberSuffixFlags.BigInteger:
+                            result = bigValue;
+                            return true;
+                        case NumberSuffixFlags.None:
+                            // Type not specified; fit value into narrowest signed type available, int32 minimum
+                            if (Utils.TryCast(bigValue, out int i))
                             {
-                                result = b;
+                                result = i;
                                 return true;
                             }
 
-                            result = null;
-                            return false;
-                        default:
-                            result = null;
-                            return false;
-                    }
-
-                    // From here on - the user hasn't specified the type, so we need to figure it out.
-                    BigInteger bigValue;
-                    TypeCode whichTryParseWorked;
-                    if (int.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out int intValue))
-                    {
-                        whichTryParseWorked = TypeCode.Int32;
-                        bigValue = intValue;
-                    }
-                    else if (long.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out long longValue))
-                    {
-                        whichTryParseWorked = TypeCode.Int64;
-                        bigValue = longValue;
-                    }
-                    else if (decimal.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out decimal decimalValue))
-                    {
-                        whichTryParseWorked = TypeCode.Decimal;
-                        bigValue = (BigInteger)decimalValue;
-                    }
-                    else
-                    {
-                        // The result must be double if we get here.
-                        if (!hex)
-                        {
-                            if (double.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out double dbl))
+                            if (Utils.TryCast(bigValue, out long l_))
```

Comment:
Could we replace the underscope with something more readable?

Reply:
Sorted!